pkg/controller/volume/pvcprotection/pvc_protection_controller.go

```diff
@@ -418,3 +505,63 @@ func (c *Controller) enqueuePVCs(pod *v1.Pod, deleted bool) {
 		}
 	}
 }
+
+func (c *Controller) getPV(pvc *v1.PersistentVolumeClaim) (*v1.PersistentVolume, error) {
+	pv, err := c.pvLister.Get(pvc.Spec.VolumeName)
+	if err != nil {
+		klog.V(3).Infof("unexpected error getting persistent volume %q from cache for claim %s/%s: %v", pvc.Spec.VolumeName, pvc.Namespace, pvc.Name, err)
+		pv, err = c.client.CoreV1().PersistentVolumes().Get(context.TODO(), pvc.Spec.VolumeName, metav1.GetOptions{})
+	}
+	if pv == nil {
+		return nil, fmt.Errorf("Empty pv object found for claim %s/%s", pvc.Namespace, pvc.Name)
+	}
+	return pv, err
+}
+
+func (c *Controller) askNodeInformer(volumeHandle string, pvc *v1.PersistentVolumeClaim) (bool, error) {
+	klog.V(3).Infof("For claim %s/%s checking for volume handle %s in node informer cache", pvc.Namespace, pvc.Name, volumeHandle)
+	if volumeHandle == "" {
+		return false, fmt.Errorf("Empty volume handle for claim %s/%s", pvc.Namespace, pvc.Name)
+	}
+
+	nodes, err := c.nodeLister.List(labels.Everything())
+	if err != nil {
+		return false, fmt.Errorf("unexpected error listing nodes from node informer: %v", err)
+	}
+	return c.isVolumeInUse(volumeHandle, nodes), nil
+}
+
+func (c *Controller) askAPIServerForNode(volumeHandle string, pvc *v1.PersistentVolumeClaim) (bool, error) {
+	klog.V(3).Infof("For claim %s/%s, volume %s, list nodes from API server", pvc.Namespace, pvc.Name, volumeHandle)
+	nodes, err := c.client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return false, fmt.Errorf("List of nodes from API server failed: %s", err.Error())
+	}
+	nodesPtr := []*v1.Node{}
+	for i := range nodes.Items {
+		nodesPtr = append(nodesPtr, &nodes.Items[i])
+	}
+
+	return c.isVolumeInUse(volumeHandle, nodesPtr), nil
+}
+
+// Helper function to determine if a given volume is present in the node's volumesInUse list.
+func (c *Controller) isVolumeInUse(volHandle string, nodes []*v1.Node) bool {
+	for _, node := range nodes {
+		for _, volInUse := range node.Status.VolumesInUse {
+			volInUse = v1.UniqueVolumeName(strings.TrimPrefix(string(volInUse), volumeHandleCSIPrefix))
```

Comment:
UniqueVolumeName is needed?

Reply:
not strictly needed. I was re-using the volInUse variable. Storing as a string now in a new variable.